### PR TITLE
[Diffusion][NPU] Add support for MOVA

### PR DESCRIPTION
**Pull request by LLThomas**
## Motivation

This PR adds NPU support to the MOVA model.

ref: [Roadmap](https://github.com/sgl-project/sglang/issues/18967)

## Modifications

1. Fix aclnnCat calling failure on NPU backend.
2. Resolve channels_last_3d incompatibility for contiguous() on NPU.
3. Optimize RoPE fallback with torch_npu.npu_rotary_mul.

### NPU (910B)
CMD: 
python -m sglang.multimodal_gen.runtime.entrypoints.cli.main generate --model-path=/path/to/MOVA-360p --prompt="A beautiful sunset over the ocean" --image-path="/path/to/sunset.jpg" --num-gpus 2 --tp-size 2

* Before:
Execution failed

* After:

https://github.com/user-attachments/assets/133cec91-9417-4c12-82df-670427b4c44a

### GPU (A100)

CMD: 
same to NPU

* Before:

https://github.com/user-attachments/assets/d25640e8-ee8f-4818-adb2-387493ee6ee4

* After:

https://github.com/user-attachments/assets/748dd252-c14c-41c8-a58a-00e544027063

## Speed Tests and Profiling

NPU (910B * 2)

CMD:
python -m sglang.multimodal_gen.runtime.entrypoints.cli.main generate --model-path=/path/to/MOVA-360p --prompt="A beautiful sunset over the ocean" --image-path="/path/to/sunset.jpg" --num-gpus 2 --tp-size 2 --perf-dump-path ../perf.json

<details>
<summary> perf.json details </summary>

```json
{
  "timestamp": "2026-03-28T05:33:56.351047+00:00",
  "request_id": "05c0954c-8d72-420e-95f8-54cc59a45f4d",
  "commit_hash": "2d288ba8c92f249d5d0d8c98b70751396ebf6f57",
  "tag": "cli_generate",
  "total_duration_ms": 1442029.9231852405,
  "steps": [
    {
      "name": "InputValidationStage",
      "duration_ms": 348.9822098053992
    },
    {
      "name": "TextEncodingStage",
      "duration_ms": 5167.57582500577
    },
    {
      "name": "ImageVAEEncodingStage",
      "duration_ms": 32805.15577085316
    },
    {
      "name": "MOVALatentPreparationStage",
      "duration_ms": 1.4249910600483418
    },
    {
      "name": "MOVATimestepPreparationStage",
      "duration_ms": 1.059889793395996
    },
    {
      "name": "MOVADenoisingStage",
      "duration_ms": 1394177.5044971146
    },
    {
      "name": "MOVADecodingStage",
      "duration_ms": 9497.18012707308
    }
  ],
  "denoise_steps_ms": [
    {
      "step": 0,
      "duration_ms": 33048.6817676574
    },
    {
      "step": 1,
      "duration_ms": 27540.082891937345
    },
    {
      "step": 2,
      "duration_ms": 27543.46876917407
    },
    {
      "step": 3,
      "duration_ms": 27564.08131122589
    },
    {
      "step": 4,
      "duration_ms": 27572.385392151773
    },
    {
      "step": 5,
      "duration_ms": 27563.069929834455
    },
    {
      "step": 6,
      "duration_ms": 27554.35159895569
    },
    {
      "step": 7,
      "duration_ms": 27544.700287282467
    },
    {
      "step": 8,
      "duration_ms": 27545.560833066702
    },
    {
      "step": 9,
      "duration_ms": 27560.257910750806
    },
    {
      "step": 10,
      "duration_ms": 27558.356884866953
    },
    {
      "step": 11,
      "duration_ms": 27562.871324829757
    },
    {
      "step": 12,
      "duration_ms": 27542.85261593759
    },
    {
      "step": 13,
      "duration_ms": 27534.70282582566
    },
    {
      "step": 14,
      "duration_ms": 27556.59586004913
    },
    {
      "step": 15,
      "duration_ms": 27554.725165944546
    },
    {
      "step": 16,
      "duration_ms": 27562.35018791631
    },
    {
      "step": 17,
      "duration_ms": 27558.70906310156
    },
    {
      "step": 18,
      "duration_ms": 37494.191779289395
    },
    {
      "step": 19,
      "duration_ms": 27553.44924610108
    },
    {
      "step": 20,
      "duration_ms": 27550.2935401164
    },
    {
      "step": 21,
      "duration_ms": 27549.314457923174
    },
    {
      "step": 22,
      "duration_ms": 27542.06692101434
    },
    {
      "step": 23,
      "duration_ms": 27541.823348030448
    },
    {
      "step": 24,
      "duration_ms": 27534.0299801901
    },
    {
      "step": 25,
      "duration_ms": 27542.893416248262
    },
    {
      "step": 26,
      "duration_ms": 27541.230192873627
    },
    {
      "step": 27,
      "duration_ms": 27549.92396896705
    },
    {
      "step": 28,
      "duration_ms": 27549.031677190214
    },
    {
      "step": 29,
      "duration_ms": 27543.825132306665
    },
    {
      "step": 30,
      "duration_ms": 27539.762417785823
    },
    {
      "step": 31,
      "duration_ms": 27543.440429959446
    },
    {
      "step": 32,
      "duration_ms": 27533.849772065878
    },
    {
      "step": 33,
      "duration_ms": 27530.339628923684
    },
    {
      "step": 34,
      "duration_ms": 27538.736396003515
    },
    {
      "step": 35,
      "duration_ms": 27536.152893211693
    },
    {
      "step": 36,
      "duration_ms": 27540.032375603914
    },
    {
      "step": 37,
      "duration_ms": 27532.260360661894
    },
    {
      "step": 38,
      "duration_ms": 27536.8240037933
    },
    {
      "step": 39,
      "duration_ms": 27547.218392137438
    },
    {
      "step": 40,
      "duration_ms": 27538.26627600938
    },
    {
      "step": 41,
      "duration_ms": 27534.237804356962
    },
    {
      "step": 42,
      "duration_ms": 27530.491921119392
    },
    {
      "step": 43,
      "duration_ms": 27540.967030916363
    },
    {
      "step": 44,
      "duration_ms": 27550.79106008634
    },
    {
      "step": 45,
      "duration_ms": 27536.134408786893
    },
    {
      "step": 46,
      "duration_ms": 27528.610394801944
    },
    {
      "step": 47,
      "duration_ms": 27525.37116408348
    },
    {
      "step": 48,
      "duration_ms": 27526.621586177498
    },
    {
      "step": 49,
      "duration_ms": 27530.62317101285
    }
  ],
  "memory_checkpoints": {
    "before_forward": {
      "allocated_mb": 0.0,
      "reserved_mb": 0.0,
      "peak_allocated_mb": 0.0,
      "peak_reserved_mb": 0.0
    },
    "after_forward": {
      "allocated_mb": 34421.4,
      "reserved_mb": 51406.0,
      "peak_allocated_mb": 39303.89,
      "peak_reserved_mb": 51406.0
    }
  },
  "meta": {
    "prompt": [
      "A beautiful sunset over the ocean"
    ],
    "model": "/path/to/openmoss/MOVA-360p"
  }
}
```
</details>

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

### CODE REVIEW

File: python/sglang/multimodal_gen/runtime/layers/rotary_embedding/utils.py [hunk 1]
```diff
@@ -111,8 +113,15 @@ def apply_flashinfer_rope_qk_inplace(
...
-        q_rot = apply_rotary_embedding(q_flat, cos, sin, interleaved=not is_neox)
-        k_rot = apply_rotary_embedding(k_flat, cos, sin, interleaved=not is_neox)
+
+        if _is_npu and is_neox:
+            cos = torch.cat([cos, cos], dim=-1)
+            sin = torch.cat([sin, sin], dim=-1)
+            q_rot, k_rot = apply_rotary_pos_emb(q_flat, k_flat, cos, sin)
+        else:
+            q_rot = apply_rotary_embedding(q_flat, cos, sin, interleaved=not is_neox)
+            k_rot = apply_rotary_embedding(k_flat, cos, sin, interleaved=not is_neox)
```
Comment (ping1jing2) [inline python/sglang/multimodal_gen/runtime/layers/rotary_embedding/utils.py]: we should try to reduce `is_npu` in the codebase, as for this code, maybe you can make the change into `python\sglang\jit_kernel\diffusion\triton\npu_fallback.py`

File: python/sglang/multimodal_gen/runtime/models/dits/mova_video_dit.py [hunk 2]
```diff
@@ -522,7 +523,10 @@ def patchify(
         self, x: torch.Tensor, control_camera_latents_input: torch.Tensor | None = None
     ):
         # NOTE(dhyu): avoid slow_conv
-        x = x.contiguous(memory_format=torch.channels_last_3d)
+        if current_platform.is_npu:
+            x = x.contiguous()
```
Comment (Makcum888e) [inline python/sglang/multimodal_gen/runtime/models/dits/mova_video_dit.py:527]: please, add comment `torch.channels_last_3d not supported on NPU`

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/model_specific_stages/mova.py [hunk 3]
```diff
@@ -714,7 +715,12 @@ def inference_single_step(
 
         # Build visual freqs for full sequence
         visual_dit._init_freqs()
-        visual_freqs = tuple(freq.to(visual_x.device) for freq in visual_dit.freqs)
+        if _is_npu:
+            visual_freqs = tuple(
```
Comment (Makcum888e) [inline python/sglang/multimodal_gen/runtime/pipelines_core/stages/model_specific_stages/mova.py:720]: please, add comment `remove it  when torch.cat on NPU is supported torch.complex128`

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/model_specific_stages/mova.py [hunk 4]
```diff
@@ -734,17 +740,22 @@ def inference_single_step(
 
         # Build audio freqs for full sequence
         self.audio_dit._init_freqs()
+        if _is_npu:    
+            audio_freqs = tuple(
```
Comment (Makcum888e) [inline python/sglang/multimodal_gen/runtime/pipelines_core/stages/model_specific_stages/mova.py:747]: please, add comment `remove it  when torch.cat on NPU is supported torch.complex128`

File: python/sglang/multimodal_gen/runtime/layers/rotary_embedding/utils.py [hunk 5]
```diff
@@ -111,8 +113,15 @@ def apply_flashinfer_rope_qk_inplace(
             sin = cos_sin_cache[positions, half_size:].to(q.dtype)
         q_flat = q.reshape(bsz * seqlen, nheads, d)
         k_flat = k.reshape(bsz * seqlen, nheads, d)
-        q_rot = apply_rotary_embedding(q_flat, cos, sin, interleaved=not is_neox)
-        k_rot = apply_rotary_embedding(k_flat, cos, sin, interleaved=not is_neox)
+
+        if _is_npu and is_neox:
+            cos = torch.cat([cos, cos], dim=-1)
```
Comment (Makcum888e) [inline python/sglang/multimodal_gen/runtime/layers/rotary_embedding/utils.py]: can we modify `apply_rotary_embedding_native` in `python/sglang/jit_kernel/diffusion/triton/npu_fallback.py‎` instead of adding `is_npu` here?

File: python/sglang/multimodal_gen/runtime/models/dits/mova_video_dit.py [hunk 6]
```diff
@@ -522,7 +523,10 @@ def patchify(
...
+        else:
+            x = x.contiguous(memory_format=torch.channels_last_3d)
```
Comment (ping1jing2) [inline python/sglang/multimodal_gen/runtime/models/dits/mova_video_dit.py]: please ensure comments are consistent with the code
```python
# NOTE(dhyu): avoid slow_conv
x = x.contiguous(memory_format=torch.channels_last_3d)
```

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/model_specific_stages/mova.py (hunk 3 above)
Comment (LLThomas) [reply python/sglang/multimodal_gen/runtime/pipelines_core/stages/model_specific_stages/mova.py:720]: done

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/model_specific_stages/mova.py (hunk 4 above)
Comment (LLThomas) [reply python/sglang/multimodal_gen/runtime/pipelines_core/stages/model_specific_stages/mova.py:747]: done

File: python/sglang/multimodal_gen/runtime/layers/rotary_embedding/utils.py (hunk 5 above)
Comment (LLThomas) [reply python/sglang/multimodal_gen/runtime/layers/rotary_embedding/utils.py]: done

File: python/sglang/multimodal_gen/runtime/layers/rotary_embedding/utils.py (hunk 1 above)
Comment (LLThomas) [reply python/sglang/multimodal_gen/runtime/layers/rotary_embedding/utils.py]: done

File: python/sglang/multimodal_gen/runtime/models/dits/mova_video_dit.py (hunk 6 above)
Comment (LLThomas) [reply python/sglang/multimodal_gen/runtime/models/dits/mova_video_dit.py]: done

**Review by Makcum888e — APPROVED** [review]
LGTM

**Review by ping1jing2 — APPROVED** [review]

**Comment by ping1jing2** [comment]
I merged it as all Diffusion CIs passed, and we will try to optimize the CI workflow in the next PR
